### PR TITLE
MLPAB-1030, MLPAB-1031: Handle fee denied event

### PR DESCRIPTION
**Pull request by acsauk**
# Purpose

Handles fee-denied events and ensures the amount to pay is correct.

I've replaced our modified open API spec for GOV UK Pay with the version from their repo now they've corrected the date format mistakes. To make the local version of the app work correctly I've had to add some named examples to the spec as we build the docker container that we reference in the mock server. This felt like a halfway house between keeping a local version of the spec and it getting out of date vs. taking the vanilla spec and then having the fee calculation act weirdly locally.

Fixes MLPAB-1030, MLPAB-1031

### CODE REVIEW

File: internal/templatefn/fn.go [hunk 1]
```diff
@@ -375,3 +377,7 @@ func concatOr(app page.AppData, list []string) string {
 func concatComma(list []string) string {
 	return strings.Join(list, ", ")
 }
+
+func penceToPounds(pence int) string {
+	return humanize.CommafWithDigits(float64(pence)/100, 2)
```
Comment (acsauk) [inline internal/templatefn/fn.go:382]: I was going a bit crazy trying to roll my own version of this but happy for any alternative on the PR 👍

Comment (hawx) [reply internal/templatefn/fn.go:382]: I'm fine with this, but hope we'll never need commas

File: .codecov.yml [hunk 2]
```diff
@@ -11,9 +11,9 @@ coverage:
     - "./internal/page/fixtures.go"
     - "./internal/page/testing_start.go"
     - "./cmd/mlpa/main.go"
-    - "./cmd/GOVUKNotify/main.go"
-    - "./cmd/GOVUKSignIn/main.go"
-    - "./cmd/OrdnanceSurveyPlacesAPI/main.go"
```
Comment (hawx) [inline .codecov.yml:16]: of course I'd miss somewhere

File: cmd/event-received/main.go [hunk 3]
```diff
@@ -35,6 +35,10 @@ type moreEvidenceRequiredEvent struct {
 	UID string `json:"uid"`
 }
 
+type feeDeniedEvent struct {
+	UID string `json:"uid"`
+}
```
Comment (hawx) [inline cmd/event-received/main.go:40]: maybe we should have had one `uidEvent`

File: docker/mock-pay/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,10 @@
+
+FROM badouralix/curl-jq:latest as build
+
+RUN curl https://raw.githubusercontent.com/alphagov/pay-publicapi/master/openapi/publicapi_spec.json --output openapi-spec.json
+RUN cat openapi-spec.json | jq '.paths."/v1/payments/{paymentId}".get.responses."200".content."application/json".examples = { "full-fee": { value: { amount: 8200, payment_id: "hu20sqlact5260q2nanm0q8u93", reference: "your-reference" } }, "half-fee": { value: { amount: 4100, payment_id: "hu20sqlact5260q2nanm0q8u93", reference: "your-reference" } }}' > mktemp && mv mktemp openapi-spec.json
```
Comment (hawx) [inline docker/mock-pay/Dockerfile:5]: ha, this is a fun way of adding it

why `> mktemp && mv mktemp openapi-spec.json` and not `> openapi-spec.json` at the end? Or `> modified-openapi-spec.json` and then use that in the next step

File: internal/page/data.go [hunk 5]
```diff
@@ -394,6 +394,20 @@ func (l *Lpa) TrustCorporationsNames() []string {
 	return names
 }
 
+func (l *Lpa) FeeAmount() int {
+	paid := 0
+
+	for _, payment := range l.PaymentDetails {
+		paid = paid + payment.Amount
```
Comment (hawx) [inline internal/page/data.go:401]: += does exist in Go

Comment (acsauk) [reply internal/page/data.go:401]: Haha I love that I can hear the tone in this comment 😂  noted ✅

File: cmd/event-received/main.go (hunk 3 above)
Comment (acsauk) [reply cmd/event-received/main.go:40]: It is becoming quite apparent now...I was planning on holding off until the event catalogue is fully populated to see if there are any unexpected useful data points they are planning on sending over but seeing as everything working on just a UID makes sense to consolidate now 👍

File: docker/mock-pay/Dockerfile (hunk 4 above)
Comment (acsauk) [reply docker/mock-pay/Dockerfile:5]: That works too - some leftover copy-paste from SO.

File: .codecov.yml (hunk 2 above)
Comment (acsauk) [reply .codecov.yml:16]: 🤷

File: internal/templatefn/fn.go (hunk 1 above)
Comment (acsauk) [reply internal/templatefn/fn.go:382]: Things will have gone very wrong on the affordability front if so

**Review by hawx — APPROVED** [review]

**Pull request merged**